### ISSUE TITLE
inference on setitem produces incorrect typing (pandas example)

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#2087]
**Describe the bug**
Using pandas, if you create a new column pyright is inferring the type of that column as whatever it was set to, as opposed to using `__getitem__` signature to infer the type

**To Reproduce**
```python
import pandas as pd

df = pd.DataFrame({"x": [1, 2, 1], "y": [3, 4, 5]})

df["z"] = [7, 8, 9]
df["a"] = 3

reveal_type(df["z"])
reveal_type(df["a"])
reveal_type(df["x"])
```
produces
```
Type of "df["z"]" is "list[int]"
Type of "df["a"]" is "Literal[3]"
Type of "df["x"]" is "Series[Dtype@__getitem__]"
```

**Expected behavior**
```
Type of "df["z"]" is "Series[Dtype@__getitem__]"
Type of "df["a"]" is "Series[Dtype@__getitem__]"
Type of "df["x"]" is "Series[Dtype@__getitem__]"
```

**VS Code extension or command-line**
VSCode pylance 2021.7.3

**Additional context**
I don't think this is a pandas stubs problem, because the error occurs when setting a constant or a list, and in both cases, it's valid pandas code, and pandas converts the expression to a `Series`

**Comment by erictraut** [comment]
This isn't an inference issue, it's a type narrowing issue. It's currently working "as designed". We'll need to decide whether to make a change here. It's related to this, which we recently closed as "won't fix": https://github.com/microsoft/pyright/issues/1814

In general, it's a bad practice (IMO) to have asymmetric get/set interfaces where the set performs some transform of the type. Unfortunately, this practice is somewhat common in Python interfaces.

I'll need to think about this more. It is possible to fix this in the type evaluator, but only at some runtime expense that will slow down type checking in general.

**Comment by erictraut** [comment]
For all member access expressions and subscript expressions that use a string or integer literal index, pyright assumes that writing to the value will result in the same value being read at a later time. This is a valid assumption the vast majority of the time, and it's an important assumption to make for type narrowing purposes.

Pyright will not make this assumption if you use a non-literal index value. So you can work around this issue by using a variable as an index or some other expression other than a simple string literal, such as `"z"""`.

Like I mentioned above, this is technically possible to "fix", but it would require a very expensive check to be performed at analysis time, which would slow down analysis for all code. Furthermore, it would probably require disabling the type narrowing assumption for all classes that use overloaded `__getitem__` or `__setitem__` methods, which would be a functionality regression in many cases.

I'm therefore inclined to mark this as "won't fix" — at least for now. I recommend using one of the workarounds suggested above if your code is using a class that implements an asymmetric `__getitem__` / `__setitem__` and your code relies on this asymmetry for type analysis.

**Comment by hmc-cs-mdrissi** [comment]
Numpy, tensorflow, pytorch, and really any tensor library will have the exact same issue as it's normal code to write stuff like,

a[1,:] = 1

even though a[1,:] will return an array and not an int. That notation is normal for math arrays so having it change for library that is meant to be used for math heavy tasks is very unlikely and would also be one of the largest breaking changes those libraries could make as dimensions are very common source of error and can not be changed easily. Today the issue doesn't matter much as these libraries don't have great type stubs yet (plus will need variadic generics+ more stuff before it's normal), but if the tensor type checking improvements continue eventually this will be a common issue for those libraries. tensorflow currently has nearly no type hints so pretty sad there.

In practice I'm not sure how often you would setitem and then use the exact same getitem. So maybe even though setitem is common if the same getitem is rare this won't affect people much. Most likely place is pandas + matplotlib if you are doing data cleaning.

**Comment by Dr-Irv** [comment]
> I'm therefore inclined to mark this as "won't fix" — at least for now. I recommend using one of the workarounds suggested above if your code is using a class that implements an asymmetric `__getitem__` / `__setitem__` and your code relies on this asymmetry for type analysis.

I'm not sure this is being friendly to the average user.  This is happening while using the pandas library.  A user (which was the case for someone on my staff) won't understand anything about the asymmetry of `__getitem__` / `__setitem__`  and why they have to use the workaround.

**Comment by erictraut** [comment]
I agree this isn't idea. The alternative would also be unfriendly to the average user, who expects that narrowed types are retained in most cases like this. For the vast majority of classes that support indexing (e.g. `list`), it's safe to assume that narrowed types are safe to retain when used with an index expression and a literal index. We received many bug reports (or feature requests, depending on how you look at it) requesting that we retain narrowed types for index expressions. So far, you're the only one who has submitted a bug about this issue. So in terms of raw numbers of requests, if we have to pick between one or the other, I the current behavior would appear to benefit more pyright users.

Ideally, the type checker could determine when it's safe to retain narrowed types and when it's not. That way, we wouldn't need to pick one or the other. In the case of overloaded `__getitem__` or `__setitem__`, it's tricky at best and impossible at worst to make this determination. And unfortunately, the `list` class (which is the most common case where this applies) has overloaded `__getitem__` or `__setitem__` methods. So if we conservatively assume that overloaded `__getitem__` means that it's unsafe to retain narrowed types, we'd disable this support for lists, where it is actually safe. Conversely, if we assume that an overloaded `__getitem__` means that it's safe to retain narrowed types, it wouldn't help you with the pandas DataFrame case.

So, for now, I don't see a good solution to this problem. I'll keep thinking about it, but if I'm unable to come up with some solution, we'll probably need to just live with the current behavior.

**Comment by Dr-Irv** [comment]
> So, for now, I don't see a good solution to this problem. I'll keep thinking about it, but if I'm unable to come up with some solution, we'll probably need to just live with the current behavior.

Could you make it an option that can be set by a user?

**Comment by erictraut** [comment]
> Could you make it an option that can be set by a user?

Providing an option like that would violate a bunch of principles we've established around settings. Settings are used to enable or disable specific classes of checks, not change the core type analysis behaviors. Plus, we add a setting only when we have signal that it's broadly needed. We don't have such signal here. So no, that's not a solution we'd entertain, at least not at this time.

**Comment by layday** [comment]
Perhaps only built-in types like list, tuple and dict could be narrowed, which are known to have a symmetric interface?

**Comment by erictraut** [comment]
After thinking about this more, I think the suggestion from @layday probably provides the best tradeoff. I've implemented this approach, and we'll see whether we get any negative feedback from users. @Dr-Irv, this will address the specific issue you reported with pandas DataFrame. Since it is not a built-in class, assignment-based type narrowing will no longer be applied in in subscript expressions that access items in a `DataFrame`.

**Comment by erictraut** [comment]
This is now addressed in pyright 1.1.159, which I just published. It will also be included in the next release of pylance.
